staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go

```diff
@@ -0,0 +1,601 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package queueset
+
+import (
+	"math"
+	"sync"
+	"time"
+
+	"github.com/pkg/errors"
+	"k8s.io/apimachinery/pkg/util/clock"
+	fq "k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing"
+	"k8s.io/apiserver/pkg/util/flowcontrol/metrics"
+	"k8s.io/apiserver/pkg/util/shufflesharding"
+	"k8s.io/klog"
+)
+
+// queueSetFactoryImpl implements the QueueSetFactory interface
+// queueSetFactoryImpl makes QueueSet objects.
+type queueSetFactoryImpl struct {
+	// counter can be nil and is ignored in that case
+	counter fq.GoRoutineCounter
+
+	clk clock.PassiveClock
+}
+
+// NewQueueSetFactory creates a new QueueSetFactory object
+func NewQueueSetFactory(clk clock.PassiveClock, counter fq.GoRoutineCounter) fq.QueueSetFactory {
+	return &queueSetFactoryImpl{
+		counter: counter,
+		clk:     clk,
+	}
+}
+
+// NewQueueSet creates a new QueueSet object
+// There is a new QueueSet created for each priority level.
+func (qsf queueSetFactoryImpl) NewQueueSet(name string, concurrencyLimit,
+	desiredNumQueues, queueLengthLimit, handSize int,
+	requestWaitLimit time.Duration) (fq.QueueSet, error) {
```

Comment:
How about the following?
```
type QueueSetConfig struct {
    ConcurrencyLimit int
    DesiredNumQueues int
    QueueLengthLimit int
    HandSize int
    RequestWaitLimit time.Duration
}
```

Reply:
If you give a mouse a cookie, he's going to ask for good comments on it.

> // QueueSetConfig defines the configuration of a QueueSet.
> type QueueSetConfig struct {
>    // ConcurrencyLimit is the maximum number of requests of this QueueSet that may be executing at a time
>    ConcurrencyLimit int
>    // DesiredNumQueues is the number of queues that the API says should exist now; this may be smaller
>    // than the number that actually exist now.
>    DesiredNumQueues int
>    // QueueLengthLimit is the maximum number of requests that may be waiting in a given queue at a time
>    QueueLengthLimit int
>    // HandSize is a parameter of shuffle sharding.  Upon arrival of a request, a queue is chosen by randomly
>    // dealing a "hand" of this many queues and then picking one of minimum length.
>    HandSize int
>    // RequestWaitLimit is the maximum amount of time that a request may wait in a queue.
>    // If, by the end of that time, the request has not been dispatched then it is rejected.
>    RequestWaitLimit time.Duration
> }

Reply:
Added `Name` to the suggested config (w/ a comment)

```
// QueueSetConfig defines the configuration of a QueueSet.
type QueueSetConfig struct {
	// Name is used to identify a queue set, allowing for descriptive information about its intended use
	Name string
	// ConcurrencyLimit is the maximum number of requests of this QueueSet that may be executing at a time
	ConcurrencyLimit int
	// DesiredNumQueues is the number of queues that the API says should exist now; this may be smaller
	// than the number that actually exist now.
	DesiredNumQueues int
	// QueueLengthLimit is the maximum number of requests that may be waiting in a given queue at a time
	QueueLengthLimit int
	// HandSize is a parameter of shuffle sharding.  Upon arrival of a request, a queue is chosen by randomly
	// dealing a "hand" of this many queues and then picking one of minimum length.
	HandSize int
	// RequestWaitLimit is the maximum amount of time that a request may wait in a queue.
	// If, by the end of that time, the request has not been dispatched then it is rejected.
	RequestWaitLimit time.Duration
}
```

Updated params in:
interface.go
```
	NewQueueSet(config QueueSetConfig) (QueueSet, error)
```

queueset.go
```
func (qsf queueSetFactory) NewQueueSet(config fq.QueueSetConfig) (fq.QueueSet, error) {

func (qs *queueSet) SetConfiguration(config fq.QueueSetConfig) error {

func newQueueSet(config fq.QueueSetConfig, c clock.PassiveClock, counter counter.GoRoutineCounter) (*queueSet, error) {

```